sdk/servicebus/service-bus/src/connectionContext.ts

```diff
@@ -64,6 +119,46 @@ export namespace ConnectionContext {
     const connectionContext = ConnectionContextBase.create(parameters) as ConnectionContext;
     connectionContext.clientContexts = {};
 
+    let waitForDisconnectResolve: () => void;
+    let waitForDisconnectPromise: Promise<void> | undefined;
+    Object.assign<ConnectionContext, ConnectionContextMethods>(connectionContext, {
+      isConnectionClosing() {
+        // When the connection is not open, but the remote end is open,
+        // then the rhea connection is in the process of terminating.
+        return Boolean(!this.connection.isOpen() && this.connection.isRemoteOpen());
+      },
+      async readyToOpenLink() {
+        // Check that the connection isn't in the process of closing.
+        // This can happen when the idle timeout has been reached but
+        // the underlying socket is waiting to be destroyed.
+        if (this.isConnectionClosing()) {
+          // Wait for the disconnected event that indicates the underlying socket has closed.
+          await this.waitForDisconnectedEvent();
```

Comment:
Can it so happen that `isConnectionClosing()` returns true but the `disconnected` event never gets fired?

Reply:
We should always see a ‘disconnected’ event is ‘isConnectionClosing()’ returns true. I did ask about this on the rhea repo while making this change for event hubs and got confirmation then :smile:

Reply:
I have a feeling I asked the same question when you did the EH change :)